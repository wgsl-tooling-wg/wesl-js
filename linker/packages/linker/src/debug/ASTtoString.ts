import {
  AbstractElem,
  ExpressionElem,
  ModuleElem,
  TypeRefElem,
  TypeTemplateParameter,
} from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

const maxLineLength = 150;

export function astToString(elem: AbstractElem, indent = 0): string {
  const { kind, contents } = elem as ModuleElem;
  const str = new LineWrapper(indent, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (contents) {
    childStrings = contents.map(e => astToString(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

// TODO DRY with above
export function elemToString(elem: AbstractElem): string {
  const { kind } = elem as ModuleElem;
  const str = new LineWrapper(0, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }
  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  addTextFields(elem, str) ||
    addVarishFields(elem, str) ||
    addStructFields(elem, str) ||
    addStructMemberFields(elem, str) ||
    addNameFields(elem, str) ||
    addMemberRef(elem, str) ||
    addFnFields(elem, str) ||
    addAliasFields(elem, str) ||
    addAttributeFields(elem, str) ||
    addExpressionFields(elem, str) ||
    addTypeRefFields(elem, str) ||
    addSynthetic(elem, str) ||
    addImport(elem, str) ||
    addRefIdent(elem, str) ||
    addTypedDeclIdent(elem, str) ||
    addDeclIdent(elem, str);
}

function addVarishFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  const { kind } = elem;
  if (
    kind === "var" ||
    kind === "let" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    addTypedDeclIdent(elem.name, str);
    return true;
  }
}

function addTextFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
    return true;
  }
}

function addRefIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "ref") {
    str.add(" " + elem.ident.originalName);
    return true;
  }
}

function addMemberRef(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "memberRef") {
    str.add(` ${elem.name.ident.originalName}.${elem.member.name}`);
    return true;
  }
}

function addDeclIdent(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "decl") {
    const { ident } = elem;
    str.add(" %" + ident.originalName);
    return true;
  }
}

function addTypedDeclIdent(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "typeDecl") {
    const { decl, typeRef } = elem;
    str.add(" %" + decl.ident.originalName);
    if (typeRef) {
      str.add(" : " + typeRefElemToString(typeRef));
    }
    return true;
  }
}

function addAliasFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRefElemToString(typeRef));
    return true;
  }
}

function addStructFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
    return true;
  }
}

function addStructMemberFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "member") {
    const { name, typeRef, attributes } = elem;
    if (attributes) {
      str.add(" " + attributes.map(a => "@" + (a && a.name)).join(" "));
    }
    str.add(" " + name.name);
    str.add(": " + typeRefElemToString(typeRef));
    return true;
  }
}

function addImport(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "import") {
    str.add(" " + importToString(elem.imports));
    return true;
  }
}

function addSynthetic(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "synthetic") {
    str.add(` '${elem.text}'`);
    return true;
  }
}

function addAttributeFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "attribute") {
    const { name, params } = elem;
    str.add(" @" + name);
    if (params) {
      str.add("(");
      str.add(params.map(expressionToString).join(", "));
      str.add(")");
    }
    return true;
  }
}

function addNameFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "name") {
    str.add(" " + elem.name);
    return true;
  }
}

function addFnFields(elem: AbstractElem, str: LineWrapper): true | undefined {
  if (elem.kind === "fn") {
    const { name, params, returnType, fnAttributes } = elem;

    str.add(" " + name.ident.originalName);

    str.add("(");
    const paramStrs = params
      .map(
        (
          p, // TODO DRY
        ) => {
          const { name } = p;
          const { originalName } = name.decl.ident;
          const typeRef = typeRefElemToString(name.typeRef!);
          return originalName + ": " + typeRef;
        },
      )
      .join(", ");
    str.add(paramStrs);
    str.add(")");

    fnAttributes?.forEach(a => {
      str.add(" @" + a?.name);
      if (a?.params) {
        str.add("(");
        str.add(a.params.map(expressionToString).join(", "));
        str.add(")");
      }
    });

    if (returnType) {
      str.add(" -> " + typeRefElemToString(returnType));
    }

    return true;
  }
}

function addTypeRefFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "type") {
    const { name } = elem;
    const nameStr = typeof name === "string" ? name : name.originalName;
    str.add(" " + nameStr);

    if (elem.templateParams !== undefined) {
      const paramStrs = elem.templateParams
        .map(templateParamToString)
        .join(", ");
      str.add("<" + paramStrs + ">");
    }
    return true;
  }
}
function addExpressionFields(
  elem: AbstractElem,
  str: LineWrapper,
): true | undefined {
  if (elem.kind === "expression") {
    const contents = elem.contents
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return elemToString(e);
        }
      })
      .join(" ");
    str.add(" " + contents);
    return true;
  }
}

function expressionToString(elem: ExpressionElem): string {
  const contents = elem.contents
    .map(e => {
      if (e.kind === "text") {
        return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
      } else {
        return elemToString(e);
      }
    })
    .join(" ");
  return contents;
}

function templateParamToString(p: TypeTemplateParameter): string {
  if (typeof p === "string") {
    return p;
  } else if (p.kind === "type") {
    return typeRefElemToString(p);
  } else if (p.kind === "expression") {
    return expressionToString(p);
  } else {
    console.log("unknown template parameter type", p);
    return "??";
  }
}

function typeRefElemToString(elem: TypeRefElem): string {
  if (!elem) return "?type?";
  const { name } = elem;
  const nameStr = typeof name === "string" ? name : name.originalName;

  let params = "";
  if (elem.templateParams !== undefined) {
    const paramStrs = elem.templateParams.map(templateParamToString).join(", ");
    params = "<" + paramStrs + ">";
  }
  return nameStr + params;
}
