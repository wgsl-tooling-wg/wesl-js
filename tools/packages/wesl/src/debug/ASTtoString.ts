import {
  assertThat,
  assertUnreachable,
} from "../../../mini-parse/src/Assertions.ts";
import {
  AbstractElem,
  AttributeElem,
  ContainerElem,
  ExpressionElem,
  FnElem,
  StuffElem,
  TranslateTimeExpressionElem,
  TypedDeclElem,
  TypeRefElem,
  TypeTemplateParameter,
  UnknownExpressionElem,
} from "../AbstractElems.ts";
import { importToString } from "./ImportToString.ts";
import { LineWrapper } from "./LineWrapper.ts";

const maxLineLength = 150;

export function astToString(elem: AbstractElem, indent = 0): string {
  const { kind } = elem;
  const str = new LineWrapper(indent, maxLineLength);
  str.add(kind);
  addElemFields(elem, str);
  let childStrings: string[] = [];
  if ("contents" in elem) {
    childStrings = elem.contents.map(e => astToString(e, indent + 2));
  }
  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"), false);
  }

  return str.result;
}

function addElemFields(elem: AbstractElem, str: LineWrapper): void {
  const { kind } = elem;
  if (kind === "text") {
    const { srcModule, start, end } = elem;
    str.add(` '${srcModule.src.slice(start, end)}'`);
  } else if (
    kind === "var" ||
    kind === "let" ||
    kind === "gvar" ||
    kind === "const" ||
    kind === "override"
  ) {
    addTypedDeclIdent(elem.name, str);
  } else if (kind === "struct") {
    str.add(" " + elem.name.ident.originalName);
  } else if (kind === "member") {
    const { name, typeRef, attributes } = elem;
    if (attributes) {
      str.add(" " + attributes.map(a => "@" + a.name).join(" "));
    }
    str.add(" " + name.name);
    str.add(": " + typeRefElemToString(typeRef));
  } else if (kind === "name") {
    str.add(" " + elem.name);
  } else if (kind === "memberRef") {
    const { extraComponents } = elem;
    const extraText =
      extraComponents ? debugContentsToString(extraComponents) : "";
    str.add(` ${elem.name.ident.originalName}.${elem.member.name}${extraText}`);
  } else if (kind === "fn") {
    addFnFields(elem, str);
  } else if (kind === "alias") {
    const { name, typeRef } = elem;
    const prefix = name.ident.kind === "decl" ? "%" : "";
    str.add(" " + prefix + name.ident.originalName);
    str.add("=" + typeRefElemToString(typeRef));
  } else if (kind === "attribute") {
    addAttribute(elem, str);
  } else if (kind === "expression") {
    const contents = elem.contents
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return astToString(e);
        }
      })
      .join(" ");
    str.add(" " + contents);
  } else if (kind === "type") {
    const { name } = elem;
    const nameStr = typeof name === "string" ? name : name.originalName;
    str.add(" " + nameStr);

    if (elem.templateParams !== undefined) {
      const paramStrs = elem.templateParams
        .map(templateParamToString)
        .join(", ");
      str.add("<" + paramStrs + ">");
    }
  } else if (kind === "synthetic") {
    str.add(` '${elem.text}'`);
  } else if (kind === "import") {
    str.add(" " + importToString(elem.imports));
  } else if (kind === "ref") {
    str.add(" " + elem.ident.originalName);
  } else if (kind === "typeDecl") {
    addTypedDeclIdent(elem, str);
  } else if (kind === "decl") {
    const { ident } = elem;
    str.add(" %" + ident.originalName);
  } else if (kind === "assert") {
    // Nothing to do for now
  } else if (kind === "module") {
    // Ignore this kind of elem
  } else if (kind === "param") {
    // TODO: This branch shouldn't exist
  } else if (kind === "stuff") {
    // Ignore
  } else if (kind === "translate-time-expression") {
    addTranslateTimeExpression(elem, str);
  } else if (kind === "directive") {
    str.add(` ${elem.directive}${elem.arguments.join(",")}`);
  } else {
    assertUnreachable(kind);
  }
}

function addAttribute(elem: AttributeElem, str: LineWrapper) {
  const { name, params } = elem;
  str.add(" @" + name);
  if (name === "if") {
    assertThat(params !== undefined);
    assertThat(params.every(p => p.kind === "translate-time-expression"));
    str.add("(");
    str.add(params.map(v => expressionToString(v.expression)).join(", "));
    str.add(")");
  } else {
    if (params) {
      assertThat(params.every(p => p.kind === "expression"));
      str.add("(");
      str.add(params.map(unknownExpressionToString).join(", "));
      str.add(")");
    }
  }
}

function addTypedDeclIdent(elem: TypedDeclElem, str: LineWrapper) {
  const { decl, typeRef } = elem;
  str.add(" %" + decl.ident.originalName);
  if (typeRef) {
    str.add(" : " + typeRefElemToString(typeRef));
  }
}

function addFnFields(elem: FnElem, str: LineWrapper) {
  const { name, params, returnType, fnAttributes } = elem;

  str.add(" " + name.ident.originalName);

  str.add("(");
  const paramStrs = params
    .map(
      (
        p, // TODO DRY
      ) => {
        const { name } = p;
        const { originalName } = name.decl.ident;
        const typeRef = typeRefElemToString(name.typeRef!);
        return originalName + ": " + typeRef;
      },
    )
    .join(", ");
  str.add(paramStrs);
  str.add(")");

  fnAttributes?.forEach(a => addAttribute(a, str));

  if (returnType) {
    str.add(" -> " + typeRefElemToString(returnType));
  }
}

function addTranslateTimeExpression(
  elem: TranslateTimeExpressionElem,
  str: LineWrapper,
) {
  str.add(expressionToString(elem.expression));
}

function expressionToString(elem: ExpressionElem): string {
  const { kind } = elem;
  if (kind === "binary-expression") {
    return `${expressionToString(elem.left)} ${elem.operator.value} ${expressionToString(elem.right)}`;
  } else if (kind === "unary-expression") {
    return `${elem.operator.value}${expressionToString(elem.expression)}`;
  } else if (kind === "ref") {
    return elem.ident.originalName;
  } else if (kind === "literal") {
    return elem.value;
  } else if (kind === "translate-time-feature") {
    return elem.name;
  } else if (kind === "parenthesized-expression") {
    return `(${expressionToString(elem.expression)})`;
  } else if (kind === "component-expression") {
    return `${expressionToString(elem.base)}[${elem.access}]`;
  } else if (kind === "component-member-expression") {
    return `${expressionToString(elem.base)}.${elem.access}`;
  } else if (kind === "call-expression") {
    return `${elem.function.ident.originalName}(${elem.arguments.map(expressionToString).join(", ")})`;
  } else {
    assertUnreachable(kind);
  }
}

function unknownExpressionToString(elem: UnknownExpressionElem): string {
  // TODO: Temp hack while I clean up the expression parsing
  if ("contents" in elem) {
    // @ts-ignore
    const contents = elem.contents
      // @ts-ignore
      .map(e => {
        if (e.kind === "text") {
          return "'" + e.srcModule.src.slice(e.start, e.end) + "'";
        } else {
          return astToString(e);
        }
      })
      .join(" ");
    return contents;
  }
  return astToString(elem);
}

function templateParamToString(p: TypeTemplateParameter): string {
  if (typeof p === "string") {
    return p;
  } else if (p.kind === "type") {
    return typeRefElemToString(p);
  } else if (p.kind === "expression") {
    return unknownExpressionToString(p);
  } else {
    console.log("unknown template parameter type", p);
    return "??";
  }
}

function typeRefElemToString(elem: TypeRefElem): string {
  if (!elem) return "?type?";
  const { name } = elem;
  const nameStr = typeof name === "string" ? name : name.originalName;

  let params = "";
  if (elem.templateParams !== undefined) {
    const paramStrs = elem.templateParams.map(templateParamToString).join(", ");
    params = "<" + paramStrs + ">";
  }
  return nameStr + params;
}

export function debugContentsToString(elem: StuffElem): string {
  const parts = elem.contents.map(c => {
    const { kind } = c;
    if (kind === "text") {
      return c.srcModule.src.slice(c.start, c.end);
    } else if (kind === "ref") {
      return c.ident.originalName; // not using the mapped to decl name, so this can be used for debug..
    } else {
      return `?${c.kind}?`;
    }
  });
  return parts.join(" ");
}
