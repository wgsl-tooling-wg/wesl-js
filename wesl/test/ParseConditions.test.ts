import { type expect, test } from "vitest";
import { astToString } from "../debug/ASTtoString.ts";
import { parseTest } from "./TestUtil.ts";
import { assertSnapshot } from "@std/testing/snapshot";

test("parse complex condition", async (t) => {
  const ast = parseTest("@if(true || (!foo&&!!false) )\nfn a() {}");
  await assertSnapshot(t, astToString(ast.moduleElem));
});

test("@if(false) enable f16", async (t) => {
  const src = `
    @if(false) enable f16;
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

test("@if(false) const_assert true;", async (t) => {
  const src = `
    @if(false) const_assert true;
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

test("@if(true) var x = 7", async (t) => {
  const src = `
    @if(true) var x = 7; 
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

test("conditional statement", async (t) => {
  const src = `
    fn main() {
      var x = 1;
      @if(true) x = 2 ;
    }
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

test("compound statement", async (t) => {
  const src = `
    fn main() {
      @if(false) {
        let x = 1;
      }
    }
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

test("conditional local var", async (t) => {
  const src = `
    fn main() {
      @if(true) var x = 1;
    }
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

test("@if(MOBILE) const x = 1", async (t) => {
  const src = `
    @if(MOBILE) const x = 1;
  `;
  const ast = parseTest(src);
  const astString = astToString(ast.moduleElem);
  await assertSnapshot(t, astString);
});

// test("", () => {
//   const src = `
//   `;
//   const ast = parseTest(src);
//   const astString = astToString(ast.moduleElem);
//   expect(astString).toMatchInlineSnapshot('tbd');
// });
