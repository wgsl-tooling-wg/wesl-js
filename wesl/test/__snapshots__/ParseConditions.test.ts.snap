export const snapshot = {};

snapshot[`parse complex condition 1`] = `
"module
  fn a() @if
    attribute @if(true || (!foo && !!false))
    decl %a
    statement
      text '{}'"
`;

snapshot[`@if(false) enable f16 1`] = `
"module
  text '
    '
  attribute @if(false)
  directive enable f16 @if
  text '
  '"
`;

snapshot[`@if(false) const_assert true; 1`] = `
"module
  text '
    '
  assert
    attribute @if(false)
    text ' const_assert true;'
  text '
  '"
`;

snapshot[`@if(true) var x = 7 1`] = `
"module
  text '
    '
  gvar %x @if
    attribute @if(true)
    text ' var '
    typeDecl %x
      decl %x
    text ' = 7;'
  text ' 
  '"
`;

snapshot[`conditional statement 1`] = `
"module
  text '
    '
  fn main()
    decl %main
    statement
      text '{
      '
      var %x
        text 'var '
        typeDecl %x
          decl %x
        text ' = 1'
      text ';
      '
      statement @if
        attribute @if(true)
        text ' '
        ref x
        text ' = 2 ;'
      text '
    }'
  text '
  '"
`;

snapshot[`compound statement 1`] = `
"module
  text '
    '
  fn main()
    decl %main
    statement
      text '{
      '
      statement @if
        attribute @if(false)
        text ' {
        let '
        typeDecl %x
          decl %x
        text ' = 1;
      }'
      text '
    }'
  text '
  '"
`;

snapshot[`conditional local var 1`] = `
"module
  text '
    '
  fn main()
    decl %main
    statement
      text '{
      '
      statement @if
        attribute @if(true)
        text ' '
        var %x
          text 'var '
          typeDecl %x
            decl %x
          text ' = 1'
        text ';'
      text '
    }'
  text '
  '"
`;

snapshot[`@if(MOBILE) const x = 1 1`] = `
"module
  text '
    '
  const %x @if
    attribute @if(MOBILE)
    text ' const '
    typeDecl %x
      decl %x
    text ' = 1;'
  text '
  '"
`;
